src/IR/DConToFunc.hs

```diff
@@ -0,0 +1,172 @@
+  deriving Monad                        via (ReaderT ArityEnv Compiler.Pass)
+  deriving MonadFail                    via (ReaderT ArityEnv Compiler.Pass)
+  deriving (MonadError Compiler.Error)  via (ReaderT ArityEnv Compiler.Pass)
+  deriving (MonadReader ArityEnv)       via (ReaderT ArityEnv Compiler.Pass)
+
+-- | Run a computation within an arity environment
+runArityFn :: ArityFn a -> ArityEnv -> Compiler.Pass a
+runArityFn (ArityFn m) = runReaderT m
+
+-- | Create a map of (DConId, Arity) key-value pairs from a list of type defintions
+createArityEnv :: [(I.TConId, TypeDef Poly.Type)] -> ArityEnv
+createArityEnv defs = M.fromList $ concatMap arities defs
+ where
+  arities :: (I.TConId, TypeDef Poly.Type) -> [(I.DConId, Int)]
+  arities (_, TypeDef { variants = vars }) = map (second variantFields) vars
+
+{- | 'dConToFunc' modifies programDefs and traverses the IR to accomplish two tasks:
+
+  1) Add top-level constructor functions for each non-nullary 'DCon' to progamDefs
+  2) Turn non-nullary data constuctors into calls to top level constructor funcs
+-}
+dConToFunc :: I.Program Poly.Type -> Compiler.Pass (I.Program Poly.Type)
+dConToFunc p@I.Program { I.programDefs = defs, I.typeDefs = tDefs } =
+  runArityFn p' (createArityEnv tDefs)
+ where
+  p' = do
+    defs' <- everywhereM (mkM dataToApp) defs
+    return p { I.programDefs = defs' ++ concat (createFuncs <$> tDefs) }
+   where
+    createFuncs (tconid, TypeDef { variants = vars }) =
+      createFunc tconid `mapMaybe` vars
+
+{- | Replace data constructor application with function application
+
+  Turn I.App instances of the form (I.Data _ _) into (I.Var _ _)
+-}
+dataToApp :: I.Expr Poly.Type -> ArityFn (I.Expr Poly.Type)
+dataToApp a@(I.Data dconid t) = do
+  Just arity <- asks (M.lookup dconid)
+  case arity of
+    0 -> return a -- leave nullary data constructors alone
+    _ -> return $ I.Var (nameFunc dconid) t
+dataToApp a = pure a
+
+{- | Create a top level function for each data constructor
+
+  Returns Nothing for nullary data constructors, 
+  which don't need top-level constructor functions.
+-}
+createFunc
+  :: I.TConId
+  -> (I.DConId, TypeVariant Poly.Type)
+  -> Maybe (I.VarId, I.Expr Poly.Type)
+  -- case of nullary dcon; guarantees params to be non-empty in the next pattern
+createFunc _    (_     , VariantNamed []    ) = Nothing
+createFunc tcon (dconid, VariantNamed params) = Just (func_name, lambda)
+ where
+  func_name = nameFunc dconid -- distinguish func name from fully applied dcon in IR
+  lambda    = I.makeLambdaChain (first Just <$> params) body
+  body      = I.zipApp dcon args
+  dcon      = I.Data (fromId dconid) t
+  args      = zip (reverse $ uncurry I.Var <$> params) (tconTyp : ts)
+  tconTyp   = Poly.TCon tcon []
+  (t : ts) =
+    reverse $ foldl1 arrow . (++ [tconTyp]) <$> tail (inits (snd <$> params))
+  {- ^ Turns a list of types into arrow types representing nested lambdas
```

Comment:
Does this appear in haddock? If not, remove the `^`. If so, add `@` around code snippets in the comments for better markup.

Reply:
Regardless of the carrot this comment doesn't appear in haddock. Only dContoFunc appears to be documented.
![image](https://user-images.githubusercontent.com/26989314/163294400-f2a95002-3528-4bf6-b7d5-33de7c090505.png)
Is this normal? The comments are visible in the source:
![image](https://user-images.githubusercontent.com/26989314/163294573-61878911-1eb8-4a91-98e7-0274955860aa.png)

Reply:
Yeah, I guess only documentation for top-level declarations are rendered